src/test/app/ValidatorKeys_test.cpp

```diff
@@ -0,0 +1,148 @@
+        "VUSmEydzBpMjFlcTNNWXl3TFZKWm5GT3I3QzBrdzJBaVR6U0NqSXpkaXRROD0ifQ==\n"};
+
+    const std::string tokenManifest =
+        "JAAAAAFxIe1FtwmimvGtH2iCcMJqC9gVFKilGfw1/vCxHXXLplc2GnMhAkE1agqXxBwD"
+        "wDbID6OMSYuM0FDAlpAgNk8SKFn7MO2fdkcwRQIhAOngu9sAKqXYouJ+l2V0W+sAOkVB"
+        "+ZRS6PShlJAfUsXfAiBsVJGesaadOJc/aAZokS1vymGmVrlHPKWX3Yywu6in8HASQKPu"
+        "gBD67kMaRFGvmpATHlGKJdvDFlWPYy5AqDedFv5TJa2w0i21eq3MYywLVJZnFOr7C0kw"
+        "2AiTzSCjIzditQ8=";
+
+public:
+    void
+    run() override
+    {
+        beast::Journal j;
+
+        // Keys when using [validation_seed]
+        auto const seedSecretKey =
+            generateSecretKey(KeyType::secp256k1, *parseBase58<Seed>(seed));
+        auto const seedPublicKey =
+            derivePublicKey(KeyType::secp256k1, seedSecretKey);
+
+        // Keys when using [validation_token]
+        auto const tokenSecretKey = *parseBase58<SecretKey>(
+            TokenType::TOKEN_NODE_PRIVATE, tokenSecretStr);
+
+        auto const tokenPublicKey =
+            derivePublicKey(KeyType::secp256k1, tokenSecretKey);
+
+        {
+            // No config -> no key but valid
+            Config c;
+            ValidatorKeys k{c, j};
+            BEAST_EXPECT(k.publicKey.size() == 0);
+            BEAST_EXPECT(k.manifest.empty());
+            BEAST_EXPECT(!k.configInvalid());
+        }
+        {
+            // validation seed section -> empty manifest and valid seeds
+            Config c;
+            c.section(SECTION_VALIDATION_SEED).append(seed);
+
+            ValidatorKeys k{c, j};
+            BEAST_EXPECT(k.publicKey == seedPublicKey);
+            BEAST_EXPECT(k.secretKey == seedSecretKey);
+            BEAST_EXPECT(k.manifest.empty());
+            BEAST_EXPECT(!k.configInvalid());
+        }
+
+        {
+            // validation seed bad seed -> invalid
+            Config c;
+            c.section(SECTION_VALIDATION_SEED).append("badseed");
+
+            ValidatorKeys k{c, j};
+            BEAST_EXPECT(k.configInvalid());
+            BEAST_EXPECT(k.publicKey.size() == 0);
+            BEAST_EXPECT(k.manifest.empty());
+        }
+
+        {
+            // validator token
+            Config c;
+            c.section(SECTION_VALIDATOR_TOKEN).append(tokenBlob);
+            ValidatorKeys k{c, j};
+
+            BEAST_EXPECT(k.publicKey == tokenPublicKey);
+            BEAST_EXPECT(k.secretKey == tokenSecretKey);
+            BEAST_EXPECT(k.manifest == tokenManifest);
+            BEAST_EXPECT(!k.configInvalid());
+        }
+        {
+            // invalid validator token
+            Config c;
+            c.section(SECTION_VALIDATOR_TOKEN).append("badtoken");
+            ValidatorKeys k{c, j};
+            BEAST_EXPECT(k.configInvalid());
+            BEAST_EXPECT(k.publicKey.size() == 0);
+            BEAST_EXPECT(k.manifest.empty());
+        }
+
+        {
+            // Token takes precedence over seed
```

Comment:
Good question, I clearly missed that since I added this test.  From a code standpoint, it would be nice to move that check over, but I like the idea of validating as early as possible, which would be where the config currently does the work.   What would you prefer?